RFC-0039-generalize-pytorch-ut.md

```diff
@@ -0,0 +1,62 @@
+
+# [RFC] Generalization of PyTorch framework UT for non-cuda device execution
+
+**Authors:**
+* @ankurneog
+
+
+## **Summary**
+Modify PyTorch framework UTs so that non-cuda devices such as intel Gaudi and intel XPU is able to harness the content and improve quality.
+
+
+## **Motivation**
+The Pytorch framework UTs are good indicator for device stack health, however these are mostly written for cpu and cuda devices, which restricts its use for non-cuda devices.
+
+We propose to modify the content wherever possible to make it available for non-cuda device execution
+
+This will also ensure greater participation for content enhancement.
+
+### **Examples**
+
+*  The execution is blocked for non-native devices using decorators such as ```onlyNativeDevices```
+*  The execution is blocked for cuda only using decorators such as ```onlyNCCL``` or ```onlyCUDA```
+*  Need scalable mechanism to select Dtypes per op described in OpInfo or ModuleInfo instead of using separate variable similar to ```dtypesIfCUDA```
+*  Need scalable mechanism to skip for different devices instead of using specific decorator ```skipIfCUDA```
+*  The dynamo content should be refactored to allow tweaking per platform/device for eg. addition of custom backends or skipping in case of unsupported backends
+*  Distributed content assumes most execution is done for nccl and gloo, with almost entire non-cpu content hard coded for nccl.
+
+## **Proposed Implementation**
+Since the content is huge, we propose a staggered approach for the implementation
+Steps:
+*   Remove restriction imposed through @onlyNativeDevices in core content, replace these with hooks so that supported devices can enable their content selectively.
+These should be flexible enough to support both in-tree and out-of-tree devices.
+*   Dtypes for a device should be dynamically loaded per op based on a common dictionary, instead of using different variables per device , eg: dtypesIfCuda
+*   Miscelleneous decorators such as @skipIfCuda should be generalized @skipIfDevice
+*   Extend use of instantiate_device_type for all content, so that developers are forced to use generalized device code rather than using "cuda" or "cpu"
+*   Generalize common distributed content , so that it can be extended for non nccl backends such as intel's hccl and ccl
+*   Generalize the dynamo content for specific backends which other devices might want to verify with existing content, the backends should always be extracted from
+    a list that is abstracted out and the list can be appended per device per TC.
```

Comment:
This is a good list of items to generalize the test cases. Does the proposal just focus on the devices having dedicated device tags installed in PyTorch core or also support the PrivateUse1 device which is used to extend PyTorch with any out of the tree devices?

Reply:
@jgong5  : Thanks for your comment, Since i investigated mostly in lines to support intel Gaudi, which  has dedicated device tag, i have not checked the impact or support needed for PrivateUse1 devices

Reply:
Perhaps @FFFrog @Yikun have more thoughts on this?

Reply:
@jgong5 @ankurneog Sorry for the late reply.

In theory, in the test framework of PyTorch, dedicated keys are almost the same as public keys (PrivateUse1), and PrivateUse1 is already supported in the test framework.

First of all, I can`t agree more with this proposal, because Ascend NPU is currently facing the above-described problems; The solution proposed by @ankurneog can solve most of the problems we have encountered; We are currently sorting out all the problems encountered, and will add them to this RFC later, and hope that the new stuff we will add will help the proposal be more complete.

By the way, if possible, we can work together to complete this proposal and make it land in PyTorch :D